### PR TITLE
refactor: make code cleaner and safer

### PR DESCRIPTION
**Pull request by Berrysoft**
* Move some code to `BufRingMmap`.
* Extend the ring size to power of 2 implicitly.
* Avoid dangling with `ManuallyDrop`.
* Use slices instead of pointers.

By the way, I think the method `release` is not a good design. Better to add a reference of io-uring in the `BufRing` struct, and release it on drop.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -11,31 +12,64 @@ use rustix::mm::{mmap_anonymous, munmap, MapFlags, ProtFlags};
...
+        buf_ring_entry.set_bid(bid);
     }
 
-    fn as_ptr(&self) -> *mut BufRingEntry {
-        self.ptr.as_ptr()
+    /// # Safety
+    ///
+    /// The advanced count should be equal to the added count.
+    unsafe fn advance_buffers(&self, count: u16) {
+        self.atomic_tail().fetch_add(count, Ordering::Release);
```
Comment (Berrysoft) [inline src/lib.rs:77]: Do we have to `& mask` here?

Comment (Sherlock-Holo) [reply src/lib.rs:77]: https://github.com/axboe/liburing/blob/7b3245583069bd481190c9da18f22e9fc8c3a805/src/include/liburing.h#L1496

`fetch_add` is enough

File: src/lib.rs [hunk 2]
```diff
@@ -134,49 +160,48 @@ impl IoUringBufRing {
...
-            buffer.buf.as_mut(),
-            buffer.buf.capacity(),
-            buffer.id,
-            self.mask(),
-            0,
-        );
-
-        self.advance_buffers(1);
+    /// * Caller must make sure release valid buffer
+    /// * Callers should be on the same thread.
```
Comment (Sherlock-Holo) [inline src/lib.rs]: > on the same thread

I think this is unnecessary:

1. the `IoUringBufRing` is not `Sync` or `Send` (maybe we can make it `Send` in the future)
2. `IoUring` is `Send`, there is no problem to send the ring and the buf-ring to another thread

Comment (Berrysoft) [reply src/lib.rs]: Oh yes. Of course callers are on the same thread. It is necessary to avoid data race.

File: src/lib.rs [hunk 3]
```diff
@@ -188,57 +213,24 @@ impl IoUringBufRing {
...
-
-    fn atomic_tail(&self) -> &AtomicU16 {
-        // Safety: no one read/write tail ptr without atomic operation after init
-        unsafe { AtomicU16::from_ptr(BufRingEntry::tail(self.buf_ring_mmap.as_ptr()).cast_mut()) }
-    }
-
     fn mask(&self) -> u16 {
         // Safety: we just get the bufs len to calculate the mask
-        unsafe { (*self.bufs.get()).len() as u16 - 1 }
+        unsafe { (*self.bufs.get()).len().next_power_of_two() as u16 - 1 }
```
Comment (Sherlock-Holo) [inline src/lib.rs]: I think

```rust
self.buf_ring_mmap.len as _
```

is better

**Review by Sherlock-Holo — APPROVED** [review]

**Pull request merged**